### PR TITLE
Core - Fix grave not found causing crash

### PR DESCRIPTION
**Pull request by BarbzYHOOL**
This fix was provided by @FALL1N1 (can't add author commit directly on github)

**Changes proposed:**

-  Prevent crash when a grave is not found  

**Issues addressed:**

Closes https://github.com/azerothcore/azerothcore-wotlk/issues/963

**Tests performed:**

None

**How to test the changes:**

Not sure how because it's not my issue and not my fix.

**Known issues and TODO list:**

- [ ] Test the original issue
- [ ] Test the fix

### CODE REVIEW

File: src/server/game/Handlers/MovementHandler.cpp [hunk 1]
```diff
@@ -487,11 +487,11 @@ void WorldSession::HandleMovementOpcodes(WorldPacket & recvData)
                 }
                 else if (!plrMover->HasFlag(PLAYER_FLAGS, PLAYER_FLAGS_IS_OUT_OF_BOUNDS))
                 {
-                    WorldSafeLocsEntry const* grave = sObjectMgr->GetClosestGraveyard(plrMover->GetPositionX(), plrMover->GetPositionY(), plrMover->GetPositionZ(), plrMover->GetMapId(), plrMover->GetTeamId());
-
-                    if ( grave)
+                    if(WorldSafeLocsEntry const* grave = sObjectMgr->GetClosestGraveyard(plrMover->GetPositionX(), plrMover->GetPositionY(), plrMover->GetPositionZ(), plrMover->GetMapId(), plrMover->GetTeamId()))
+                    {
                         plrMover->TeleportTo(grave->map_id, grave->x, grave->y, grave->z, plrMover->GetOrientation());
                     plrMover->Relocate(grave->x, grave->y, grave->z, plrMover->GetOrientation());
```
Comment (FALL1N1) [inline src/server/game/Handlers/MovementHandler.cpp]: Space

**Comment by ShinDarth** [comment]
I don't really understand what it's actually fixing, it looks more a codestyle change to me.

File: src/server/game/Handlers/MovementHandler.cpp [hunk 2]
```diff
@@ -487,11 +487,11 @@ void WorldSession::HandleMovementOpcodes(WorldPacket & recvData)
                 }
                 else if (!plrMover->HasFlag(PLAYER_FLAGS, PLAYER_FLAGS_IS_OUT_OF_BOUNDS))
                 {
-                    WorldSafeLocsEntry const* grave = sObjectMgr->GetClosestGraveyard(plrMover->GetPositionX(), plrMover->GetPositionY(), plrMover->GetPositionZ(), plrMover->GetMapId(), plrMover->GetTeamId());
-
-                    if ( grave)
+                    if(WorldSafeLocsEntry const* grave = sObjectMgr->GetClosestGraveyard(plrMover->GetPositionX(), plrMover->GetPositionY(), plrMover->GetPositionZ(), plrMover->GetMapId(), plrMover->GetTeamId()))
```
Comment (ShinDarth) [inline src/server/game/Handlers/MovementHandler.cpp]: missing space after `if`

**Comment by FALL1N1** [comment]
@ShinDarth it will crash on the 2nd line of "grave->" if it's not found from the first declaration WorldSafeLocsEntry const* grave = ...

**Comment by BarbzYHOOL** [comment]
I thought it was because of the `WorldSafeLocsEntry const* grave`

**Comment by talamortis** [comment]
could anyone please give the cords of a GPS? oh which this happens so I could test??

**Comment by talamortis** [comment]
Ok, Now looking at the code again I understand it :P I should of not had a quick glance but why did you remove the `if (grave)`

**Comment by FALL1N1** [comment]
Because now we define the 'grave' inside the if check, saves space and it looks better

**Comment by FALL1N1** [comment]
how to reproduce: purge the graveyard table

**Comment by MajorDOT** [comment]
@FALL1N1 do you mean purge the Game_Graveyard_Zone?
I purged it, but when I die I just get teleported to the Barrens (did not apply patch) I don't crash.
I'll test in a few more areas, but I am not getting any crash.

**Comment by FALL1N1** [comment]
http://prntscr.com/k70ill check the conditions

**Comment by FrancescoBorzi** [comment]
what is the status of this?

**Comment by BarbzYHOOL** [comment]
Waiting for some good input

**Comment by wetbrownsauce** [comment]
Is there a reliable way to reproduce the crash? I've tried dropping all data from `game_graveyard_zone` and dying in various locations, both in world, instance, and bg, but I just get sent to the default barrens gy... no crash.

**Comment by wetbrownsauce** [comment]
From what I understand (I don't know much c++), the code will theoretically execute like this without the PR:


```
//get grave
WorldSafeLocsEntry const* grave = sObjectMgr->GetClosestGraveyard(plrMover->GetPositionX(), plrMover->GetPositionY(), plrMover->GetPositionZ(), plrMover->GetMapId(), plrMover->GetTeamId());

//if grave found, move player to it
if ( grave)
    plrMover->TeleportTo(grave->map_id, grave->x, grave->y, grave->z, plrMover->GetOrientation());
plrMover->Relocate(grave->x, grave->y, grave->z, plrMover->GetOrientation());
```

However, the last line in that block is outside the indentation block of the `if ( grave)` meaning it will execute even if the `if ( grave)` check fails. This results in the last line executing on a null value of grave.

The PR appears to change the code style around a bit, unintentionally masking the true effect of indenting that last line and putting {} around the two

**Comment by BarbzYHOOL** [comment]
hmmm then need to unindent the last line?

**Comment by FrancescoBorzi** [comment]
guys the indentation does not affect the behaviour in C++, I don't see how an indentation change can resolve a crash.

I would close this.

**Comment by BarbzYHOOL** [comment]
because

```
if blabla
    do this
do something else
```
means
```
if blabla
{
    do this
}
do something else
```

but the PR does this
```
if blabla
{
    do this
    do something else
}
```

**Comment by FrancescoBorzi** [comment]
@BarbzYHOOL oh yeah, my bad. I was only focussing on the first part and I assumed that the second part was indentation-change only (but it isn't).

Then I would revert the first part and keep the second one.

**Comment by Nefertumm** [comment]
It's not an indentation change. If grave was found or not, relocate will be executed anyway, causing a segmentation fault because grave will be nullptr and you're trying to dereference it.

**Comment by FALL1N1** [comment]
Exactly as @wetbrownsauce said, the grave pointer was outside the if() scope so if the grave was not found the pointer would be null hence it'd crash / cause the segmentation fault.

File: src/server/game/Handlers/MovementHandler.cpp (hunk 2 above)
Comment (FrancescoBorzi) [reply src/server/game/Handlers/MovementHandler.cpp]: this can be reverted

**Comment by Nefertumm** [comment]
It's not something related to graveyards. It's something like player out of bounds and falling.
https://github.com/azerothcore/azerothcore-wotlk/blob/eb2708bc79f62787d05dff6cd3e4ffae04213290/src/server/game/Handlers/MovementHandler.cpp#L477 Maybe @FALL1N1 can provides more information about how this crash happened to him the first time.

**Comment by FALL1N1** [comment]
@Nefertumm I have written it from the crash logs provided somewhere here in the issue tracker, I don't use ac so I can't reproduce it, sorry.

**Comment by Nefertumm** [comment]
According to the if conditions in the code, this might be tested with the following instructions:
- Drop the game_graveyard_zone table data.
- DIE and release the spirit.
- Go to any map.
- Fall off of that map (Outland might be a good option for this) or go with go xyz command with a very low Z coordinate (remember you can still use commands while death with /g).

Remember not to use gm fly while testing or gm on.

**Comment by wetbrownsauce** [comment]
Tried, no crash. gm off and gm fly off

- jumped off edge of nagrand as ghost: barrens gy
- .go xyz'd to -1000 z in borean tundra both as ghost and while living: barrens gy
- .go xyz'd to -10000 z in borean tundra both as ghost and while living: barrens gy
- .go xyz'd to -10000 z at the edge of the veiled sea fatigue zone (by the blue barriers, literal edge of the map): barrens gy

I'm at a loss of how to reproduce this, let alone how a player would ever run into this

**Comment by Nefertumm** [comment]
Well, we can't reproduce it, but the way that was programmed it's bad. If sometimes that piece of code is executed, it will be a segmentation fault crash. I think this is not intended, they missed the brackets.
In my opinion, I would merge this.

**Comment by wetbrownsauce** [comment]
Fair enough, I agree.

File: src/server/game/Handlers/MovementHandler.cpp (hunk 2 above)
Comment (FrancescoBorzi) [reply src/server/game/Handlers/MovementHandler.cpp]: You can revert the part where you put the declaration of the variable inside the if

File: src/server/game/Handlers/MovementHandler.cpp [hunk 3]
```diff
@@ -487,7 +487,9 @@ void WorldSession::HandleMovementOpcodes(WorldPacket & recvData)
                 }
                 else if (!plrMover->HasFlag(PLAYER_FLAGS, PLAYER_FLAGS_IS_OUT_OF_BOUNDS))
                 {
-                    if(WorldSafeLocsEntry const* grave = sObjectMgr->GetClosestGraveyard(plrMover->GetPositionX(), plrMover->GetPositionY(), plrMover->GetPositionZ(), plrMover->GetMapId(), plrMover->GetTeamId()))
+                    WorldSafeLocsEntry const* grave = sObjectMgr->GetClosestGraveyard(plrMover->GetPositionX(), plrMover->GetPositionY(), plrMover->GetPositionZ(), plrMover->GetMapId(), plrMover->GetTeamId());
+                    
+                    if(grave)
```
Comment (FrancescoBorzi) [inline src/server/game/Handlers/MovementHandler.cpp]: missing space after if. It should be: `if (grave)`

Comment (Nefertumm) [reply src/server/game/Handlers/MovementHandler.cpp]: Done

**Comment by FrancescoBorzi** [comment]
Feel free to merge

**Comment by Nefertumm** [comment]
I'm merging this, it can fix a potential crash. Thank you all ppl!

**Comment by BarbzYHOOL** [comment]
good job all

**Pull request merged**